src/Symfony/Component/Messenger/Adapter/PhpAmqp/AmqpReceiver.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Adapter\PhpAmqp;
+
+use Symfony\Component\Messenger\Adapter\PhpAmqp\Exception\RejectMessageExceptionInterface;
+use Symfony\Component\Messenger\Transport\ReceiverInterface;
+use Symfony\Component\Messenger\Transport\Serialization\DecoderInterface;
+
+/**
+ * Symfony Messenger receiver to get messages from AMQP brokers using PHP's AMQP extension.
+ *
+ * @author Samuel Roze <samuel.roze@gmail.com>
+ */
+class AmqpReceiver implements ReceiverInterface
+{
+    public const STOP = 'stop';
+
+    private $messageDecoder;
+    private $connection;
+
+    public function __construct(DecoderInterface $messageDecoder, Connection $connection)
+    {
+        $this->messageDecoder = $messageDecoder;
+        $this->connection = $connection;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function receive(): iterable
+    {
+        $shouldStop = false;
+
+        while (!$shouldStop) {
+            if (null === $message = $this->connection->waitAndGet()) {
+                continue;
+            }
+
+            try {
+                $shouldStop = self::STOP === (yield $this->messageDecoder->decode(array(
```

Comment:
Is this statement correct? it seems that `(yield ...)` will be `null` always https://3v4l.org/51jhv.

Could we do this `yield $shouldStop = $this->messageDecoder->decode(...)` and check it below? ... or throw a new `StopReceiveException` instead?

Reply:
Oh NVM, I see it https://github.com/symfony/symfony/pull/26632/files#diff-b1417962dd38ee12497279be6f3f1308R53 I guess we need to do that (`$g->send('stop')`) to stop the iterator here.

Reply:
So the only way to stop the iteration is via the iterator ? It could be convenient to also expose a method `stop()`. We could use this method in a listener (or something like that) that watch the memory for exemple. Or to have a nice control of signals

Reply:
Why do we have this `->send('stop')` ? The problem is that if we use a `break` to leave the `foreach` around the generator, it will not execute what is after the `yield` in this file, and therefore will not acknowledge the message. 

Having a `stop` method on the receiver is a more explicit alternative but it can't be "forwarded" by decorating generators without knowing the receiver (not that it's a massive problem as I believe all of the decorations will be decorators on the `ReceiverInterface`). 

That said, do you have any other idea on how to tackle this problem than adding a `close` method?

Reply:
> That said, do you have any other idea on how to tackle this problem than adding a close method?

Unfortunatly, I don't have a solution with the existing design :/ In my previous PR I managed to fix it by decoupling the "Loop" and the "Consumer"

Reply:
I think that `ReceiverInterface::stop` is a good idea. We wouldn't propagate the stop of the consumer back via the Generator but simply by calling `$receiver->stop()`, which would, in turn, change the `$shouldStop` boolean within the consumer. Simpler.